docs/Developer/Serialization.md

```diff
@@ -0,0 +1,373 @@
+
+- For example, if you're serializing a `std::vector<int>`, you'll need to include the following header:
+```cpp
+#include <cereal/types/vector.hpp>
+```
+
+- If you are using custom types, ensure the serialization process is correctly implemented for those types as well. The same approach applies to user-defined types, so include the appropriate headers and define the serialize function in that class or struct.
+
+<details>
+
+<summary><strong>Commonly used C++ data types and their corresponding Cereal headers</strong></summary>
+
+
+| Type                | Header to include                               |
+|:--------------------|:----------------------------------------------|
+| `std::array`        | `#include <cereal/types/array.hpp>`            |
+| `std::atomic`       | `#include <cereal/types/atomic.hpp>`           |
+| `std::bitset`       | `#include <cereal/types/bitset.hpp>`           |
+| `std::chrono`       | `#include <cereal/types/chrono.hpp>`           |
+| `std::complex`      | `#include <cereal/types/complex.hpp>`          |
+| `std::deque`        | `#include <cereal/types/deque.hpp>`            |
+| `std::forward_list` | `#include <cereal/types/forward_list.hpp>`     |
+| `std::functional`   | `#include <cereal/types/functional.hpp>`       |
+| `std::list`         | `#include <cereal/types/list.hpp>`             |
+| `std::map`          | `#include <cereal/types/map.hpp>`              |
+| `std::memory`       | `#include <cereal/types/memory.hpp>`           |
+| `std::optional`     | `#include <cereal/types/optional.hpp>`         |
+| `std::queue`        | `#include <cereal/types/queue.hpp>`            |
+| `std::set`          | `#include <cereal/types/set.hpp>`              |
+| `std::stack`        | `#include <cereal/types/stack.hpp>`            |
+| `std::string`       | `#include <cereal/types/string.hpp>`           |
+| `std::tuple`        | `#include <cereal/types/tuple.hpp>`            |
+| `std::unordered_map`| `#include <cereal/types/unordered_map.hpp>`    |
+| `std::unordered_set`| `#include <cereal/types/unordered_set.hpp>`    |
+| `std::utility`      | `#include <cereal/types/utility.hpp>`          |
+| `std::valarray`     | `#include <cereal/types/valarray.hpp>`         |
+| `std::variant`      | `#include <cereal/types/variant.hpp>`          |
+| `std::vector`       | `#include <cereal/types/vector.hpp>`           |
+
+</details>
+
+<br>
+
+### **STEP 02: ADD SERIALIZE FUNCTION**
+
+Within your class header file
+- Firstly, declare the `serialize` function inside the class:
+    - Add the following template function signature in the public section of your class to allow serialization for any archive type.
+```cpp 
+    template <class Archive> 
+    void serialize(Archive & archive); 
+```  
+- Secondly, define the `serialize` function outside the class:
+    - After your class declaration, define the serialize function at the end of the header file. 
```

Comment:
Why not in the `.cpp` file, like all the other methods?

Reply:
This is a great question to address. Included this in the note:

> Why define serialize in the header?
> Cereal relies heavily on templates, and C++ templates require full implementation details to be available during 
> compilation for proper instantiation. Since templates must be instantiated at compile time, placing the serialize 
> function in a .cpp file could result in missing template information, leading to linker errors. By defining the function 
> in the header file, the compiler has all the necessary information to properly instantiate the serialize function for 
> various data types.